cirq/optimizers/three_qubit_decomposition.py

```diff
@@ -0,0 +1,272 @@
+    Synthesis of Quantum Logic Circuits. Tech. rep. 2006,
+    https://arxiv.org/abs/quant-ph/0406176
+
+    Args:
+        q0: first qubit
+        q1: second qubit
+        q2: third qubit
+        u: unitary matrix
+        atol: A limit on the amount of absolute error introduced by the
+            construction.
+
+    Returns:
+        The resulting operations will have only known two-qubit and one-qubit
+        gates based operations, namely CZ, CNOT and rx, ry, PhasedXPow gates.
+    """
+    assert np.shape(u) == (8, 8)
+    assert cirq.is_unitary(u)
+
+    try:
+        from scipy.linalg import cossin
+    except ImportError:  # coverage: ignore
+        # coverage: ignore
+        raise ImportError(
+            "cirq.three_qubit_unitary_to_operations requires "
+            "SciPy 1.5.0+, as it uses the cossin function. Please"
+            " upgrade scipy in your environment to use this "
+            "function!"
+        )
+    (u1, u2), theta, (v1h, v2h) = cossin(u, 4, 4, separate=True)
+
+    cs_ops = _cs_to_ops(q0, q1, q2, theta)
+    if len(cs_ops) > 0 and cs_ops[-1] == cirq.CZ(q2, q0):
+        # optimization A.1 - merging the last CZ from the end of CS into UD
+        # cz = cirq.Circuit([cs_ops[-1]]).unitary()
+        # CZ(c,a) = CZ(a,c) as CZ is symmetric
+        # for the u1⊕u2 multiplexor operator:
+        # as u1(b,c) is the operator in case a = \0>,
+        # and u2(b,c) is the operator for (b,c) in case a = |1>
+        # we can represent the merge by phasing u2 with I ⊗ Z
+        u2 = u2 @ np.kron(np.eye(2), np.array([[1, 0], [0, -1]]))
+        cs_ops = cs_ops[:-1]
+
+    d_ud, c_ud = _two_qubit_multiplexor_to_circuit(q0, q1, q2, u1, u2, shift_left=True, atol=atol)
+
+    _, c_vdh = _two_qubit_multiplexor_to_circuit(
+        q0, q1, q2, v1h, v2h, shift_left=False, diagonal=d_ud, atol=atol
+    )
+
+    return list(cirq.Circuit([c_vdh, cs_ops, c_ud]).all_operations())
+
+
+def _cs_to_ops(q0: ops.Qid, q1: ops.Qid, q2: ops.Qid, theta: np.ndarray) -> List[ops.Operation]:
+    """Converts theta angles based Cosine Sine matrix to operations.
+
+    Using the optimization as per Appendix A.1, it uses CZ gates instead of
+    CNOT gates and returns a circuit that skips the terminal CZ gate.
+
+    Args:
+        a, b, c: the 3 qubits in order
+        theta: theta returned from the Cosine Sine decomposition
+    Returns:
+         the operations
+    """
+    # Note: we are using *2 as the thetas are already half angles from the
+    # CSD decomposition, but cirq.ry takes full angles.
+    angles = _multiplexed_angles(theta * 2)
+    rys = [cirq.ry(angle).on(q0) for angle in angles]
+    ops = [
+        rys[0],
+        cirq.CZ(q1, q0),
+        rys[1],
+        cirq.CZ(q2, q0),
+        rys[2],
+        cirq.CZ(q1, q0),
+        rys[3],
+        cirq.CZ(q2, q0),
+    ]
+    return _optimize_multiplexed_angles_circuit(ops)
+
+
+def _optimize_multiplexed_angles_circuit(operations: List[ops.Operation]):
```

Comment:
Yeah a general compiler that does that could be useful, but much harder task than this specific one - that's why we agreed with Craig back in the day to implement the special case here. I can open an issue though, I think it might be useful. Though a lot of cases are covered by `MergeInteractions`, this might be an interesting one to write for larger gates. 

re: `cirq.linalg.is_involutive` could be cool - we just talked about with Orion maybe introducing a `cirq.period` protocol too, where things could tell their period, which would be 2 for involutary operators.

Reply:
OK